File: scala/src/TopologicalSort.scala
import scala.collection.mutable.ArrayBuffer

object TopologicalSort {
  def dfs(graph: Array[ArrayBuffer[Int]], used: Array[Boolean], order: ArrayBuffer[Int], u: Int) {
    used(u) = true
    for (v <- graph(u); if !used(v))
      dfs(graph, used, order, v)
    order += u
  }

  def topologicalSort(graph: Array[ArrayBuffer[Int]]): Array[Int] = {
    val n: Int = graph.length
    val used = new Array[Boolean](n)
    val order = ArrayBuffer[Int]()

    for (i <- 0 until n; if !used(i))
      dfs(graph, used, order, i)

    order.reverse.toArray
  }

  // Usage example
  def main(args: Array[String]) {
    val g: Array[ArrayBuffer[Int]] = Array.fill(3)(ArrayBuffer())

    g(2) += 0
    g(2) += 1
    g(0) += 1

    val order: Array[Int] = topologicalSort(g)
    print(order.mkString(" "))
  }
}

File: kotlin/src/TopologicalSort.kt
fun topologicalSort(graph: Array<out List<Int>>): List<Int> {
    val n = graph.size
    val used = BooleanArray(n)
    val order = mutableListOf<Int>()

    fun dfs(u: Int) {
        used[u] = true
        graph[u].filter { !used[it] }.forEach { dfs(it) }
        order.add(u)
    }

    (0 until n).filter { !used[it] }.forEach { dfs(it) }
    order.reverse()
    return order
}

// Usage example
fun main(args: Array<String>) {
    val graph = (1..3).map({ mutableListOf<Int>() }).toTypedArray()
    graph[2].add(0)
    graph[2].add(1)
    graph[0].add(1)

    val order = topologicalSort(graph)
    println(order)
}
